### ISSUE TITLE
Duplicate Saga creation after switching from StructureMap

### ISSUE THREAD

**Issue by michaeldtaylor** [rebus-org/Rebus#828]
I have recently made the switch from StructureMap to Lamar to enable easier integration with .NET Core services. This basically involved replacing Rebus.StructureMap with Rebus.ServiceProvider and doing the following:

```csharp
// Configure Lamar
var serviceRegistry = new ServiceRegistry();

Startup.ConfigureContainer(serviceRegistry, AppNames.Overlord);

using (var container = new Container(serviceRegistry))
{
    ServiceLocator.Container = container;

    container.UseRebus();

    var overlordService = container.GetService<OverlordService>();

    // Configure Topshelf
    var host = HostFactory.New(x =>
    {
        x.Service<OverlordService>(sc =>
        {
            sc.ConstructUsing(s => overlordService);

            sc.WhenStarted(s => s.Start());
            sc.WhenStopped(s => s.Stop());
        });

        x.SetServiceName("Overlord");
        x.SetDisplayName("Overlord");
        x.SetDescription("The Overlord");

        x.UseNLog();

        x.RunAsLocalSystem();
        x.StartAutomatically();
    });

    return (int)host.Run();
}
```

My "Startup" bootstraps all the ServiceRegistry stuff, with my Rebus config looking like this (I have it configurable for both FileSystem and AzureServiceBus):

```csharp
services.IncludeRegistry<OrchestrationOverlordRegistry>();

// Configure Rebus
services.AddRebus(c => c
    .Options(o =>
    {
        o.EnableIdempotentSagas();
        o.SimpleRetryStrategy(secondLevelRetriesEnabled: true);
    })
    .Logging(l => l.NLog())
    .Subscriptions(s => AddrRebusConfigurer.ConfigureSubscriptions(s, configuration.Shared.Transport))
    .Routing(r => AddrRebusConfigurer.ConfigureRouting(r, configuration.Shared.Transport, QueueName, OverlordService.HandledEventTypes))
    .Transport(t => AddrRebusConfigurer.ConfigureTransport(t, configuration.Shared.Transport, QueueName))
    .Sagas(s => s.StoreInSqlServer(configuration.OverlordConnectionString, "Sagas", "SagaIndex")));
```
AddrRebusConfigurer just does a few extra things if FileSystem (like making sure the messages are mapped before I call Subscribe etc). FileSystem is only going to be used for local testing really, but its really nice to have! The messages handlers are mapped using a Lamar ServiceRegistry:

```csharp
public class OrchestrationOverlordRegistry : ServiceRegistry
{
    public OrchestrationOverlordRegistry()
    {
        Scan(x =>
        {
            x.AssemblyContainingType<OrchestrationOverlordRegistry>();
            x.WithDefaultConventions();

            x.ConnectImplementationsToTypesClosing(typeof(IApiOperationHandler<>));
            x.ConnectImplementationsToTypesClosing(typeof(IHandleMessages<>));
        });

        For<IIncomingOrderQueue>().Use<IncomingOrderAzureQueue>().Singleton();
    }
}
```
My component "OverlordService" is injected with an "IncomingQueuePoller" (Timer component) that looks at a Azure Storage Queue and translates the queue message into an "OrderReceived" bus message that it uses "SendLocal" to route to the OrderProgressSaga (which lives in Overlord). This was all working perfectly before my changes.

The problem:

Whenever the message is sent, I see the following:

Debug: Sending Domain.Events.OrderReceived -> "overlord-queue"
Debug: Created new saga data with ID 638a005d-3fd2-43b2-886d-25295d668f6c for message "OrderReceived/9613d5a8-0966-48cf-a707-3dc9ad948c89"
Debug: Created new saga data with ID 1f47a6ef-6ce6-4083-8fb0-d95e6eca898b for message "OrderReceived/9613d5a8-0966-48cf-a707-3dc9ad948c89" 

So I'm getting my OrderProgressSaga created twice with "IsNew=true" resulting in a Concurrency exception.

What I have done/checked:

1. That the IBus instance is the same for both OrderService and the IncomingQueuePoller (it is).
2. Changing between FileSystem/AzureServiceBus to see if it was a transport issue - no difference in behaviour.
3. Tried to _bus.Send() instead of _bus.SendLocal() for the IncomingQueuePoller - no difference in behvaiour (although I did have to map all messages even when configured for AzureServiceBus which was odd as that should be default for the transport?).
4. Tried not registering the IHandleMessages, just incase they were being registered twice for some reason - without the Lamar config line it tells me no handlers are registered.

Have you any idea why I might be getting this double Saga creation for a single message sent now that I'm using Lamar/ServiceProvider?

Many thanks,

Michael Taylor

**Comment by mookid8000** [comment]
If your implementation of `IHandlerActivator` returns two instances of your saga handler class, then I believe your would experience exactly these symptoms.

Could you try and call 
```csharp
var activator = <your Lamar handler activator>;

var handlers = await activator.GetHandlers<OrderReceived>();
```
and see how many instances it returns?

**Comment by michaeldtaylor** [comment]
Thanks for the response. I think you are correct, but I don't completely understand why they are being registered twice. This is probably my misunderstanding of Lamar and its closeness to StructureMap, but perhaps you have some ideas.

I don't have my own Lamar implementation of IHandlerActivator - I'm using your Rebus.ServiceProvider library (so DependencyInjectionHandlerActivator.cs).

From what I understand in this context Lamar is basically acting as an "enhanced" IServiceProvider that allows you to maintain StructureMap style registries, but allows all the .NET core registrations etc.

If you use ServiceProvider directly then according to your docs you do:

````csharp
var services = new ServiceCollection();
services.AutoRegisterHandlersFromAssemblyOf<Handler1>();

// Configure Rebus
services.AddRebus(configure => configure
    .Logging(l => l.ColoredConsole())
    .Transport(t => t.UseInMemoryTransport(new InMemNetwork(), "Messages"))
    .Routing(r => r.TypeBased().MapAssemblyOf<Message1>("Messages")));

using (var provider = services.BuildServiceProvider())
{
    provider.UseRebus();

    // Do stuff
}
````
I adapted this to create a top-level "ServiceRegistry" instead (ServiceRegistry is an IServiceCollection), add stuff to that by scanning as per StructureMap and then instead of "services.BuildServiceProvider()" I do "new Container(serviceRegistry)" (Container is a IServiceProvider).

My equivalent (flattened):

````csharp
// Configure Lamar
var serviceRegistry = new ServiceRegistry();

serviceRegistry.Scan(x =>
{
    x.AssembliesFromApplicationBaseDirectory();
    x.LookForRegistries();
});

// This bit lives in another registry that is "included" and must be the source of the problem
serviceRegistry.Scan(x =>
{
    x.AssemblyContainingType<OrchestrationOverlordRegistry>();
    x.WithDefaultConventions();

    x.ConnectImplementationsToTypesClosing(typeof(IHandleMessages<>));
});

// Configure Rebus
serviceRegistry.AddRebus(configure => configure
    .Logging(l => l.ColoredConsole())
    .Transport(t => t.UseInMemoryTransport(new InMemNetwork(), "Messages"))
    .Routing(r => r.TypeBased().MapAssemblyOf<Message1>("Messages")));

using (var container = new Container(serviceRegistry))
{
    container.UseRebus();

    // Do stuff
}
````
I believe this is a reasonable strategy? From Rebus point of view I am just using ServiceProvider, but many of the registrations were done at the Lamar level.

When I try and get the activator out of the container as per your last response I get nothing:

````csharp
var activator = container.GetService<IHandlerActivator>();

// activator = null

var orderReceived = new OrderReceived();

var handlers = await activator.GetHandlers(orderReceived, null);
````
When I inspect the container directly I can see that there are two IHandlerMessage for each message, so that is the fundamental problem.

As per my original message, if I remove the IHandlerMessage scan from the inner regsitry then no IHandlerMessages are present.

If I move the IHandlerMessage scan to the very top of the container bootstrap then only one gets registered, which is what I want:

````csharp
// Configure Lamar
var serviceRegistry = new ServiceRegistry();

serviceRegistry.Scan(x =>
{
    x.AssemblyContainingType<OrchestrationOverlordRegistry>();
    x.WithDefaultConventions();

    x.ConnectImplementationsToTypesClosing(typeof(IHandleMessages<>));
});
````

So in summary, I guess I don't understand why the Lamar ServiceRegistry scan is adding multiple IHandlerMessage instances when it is done within an included registry, which is more of a Lamar/me problem.

Thanks,

Michael Taylor

**Comment by mookid8000** [comment]
If I understood what you just wrote, you figured out how to make it work 😁 unfortunately, I've never used StructureMap or Lamar, so I'm not really familiar with what constitutes idiomatic usage of either, and I don't have any good guesses as to why multiple registrations seem to happen with your configuration.

If you figure it out, it would be awesome if you could post it in here, so it may turn up if people Bingle for issues like yours.

Thanks for the kind words btw. 🙂

**Comment by tempworks-paul** [comment]
I've been prototyping with Lamar and Rebus the last couple of months. For now I did a bit more brute force approach. I'm using the Rebus.StructureMap adapter converted to use Lamar instead and ran into this same problem. Basically I just applied a `distinct` operation on the returned MessageHandlers.

```
        public Task<IEnumerable<IHandleMessages<TMessage>>> GetHandlers<TMessage>(TMessage message, ITransactionContext transactionContext)
        {
            var scope = _provider.CreateScope();

            var resolvedHandlerInstances = GetMessageHandlersForMessage<TMessage>(scope);

            var distinctResolvedHandlerInstances =
                resolvedHandlerInstances
                    .GroupBy(g => g.GetType().FullName)
                    .Select(g => g.First());

            transactionContext.OnDisposed(scope.Dispose);

            return Task.FromResult((IEnumerable<IHandleMessages<TMessage>>)distinctResolvedHandlerInstances.ToArray());
        }
```

I've been meaning to go back and refactor but we're probably dropping Lamar in favor of MS DI w/ Scrutor.
